diff --git xbmc-13-0f3db05/configure.in xbmc-13-b8667e9/configure.in
index 2e49fa1..16253ce 100644
--- xbmc-13-0f3db05/configure.in
+++ xbmc-13-b8667e9/configure.in
@@ -692,11 +692,8 @@ case $host in
      ARCH="arm"
      use_arch="arm"
      ffmpeg_target_os=linux
-     use_joystick=no
      use_neon=yes
      use_gles=yes
-     use_sdl=no
-     use_x11=no
      use_wayland=no
      use_static_ffmpeg=yes
      ;;
diff --git xbmc-13-0f3db05/xbmc/windowing/X11/WinSystemX11GLES.cpp xbmc-13-b8667e9/xbmc/windowing/X11/WinSystemX11GLES.cpp
index b2e282e..1b46513 100644
--- xbmc-13-0f3db05/xbmc/windowing/X11/WinSystemX11GLES.cpp
+++ xbmc-13-b8667e9/xbmc/windowing/X11/WinSystemX11GLES.cpp
@@ -27,7 +27,6 @@
 #include "filesystem/SpecialProtocol.h"
 #include "settings/DisplaySettings.h"
 #include "settings/Settings.h"
-#include "settings/DisplaySettings.h"
 #include "guilib/GraphicContext.h"
 #include "guilib/Texture.h"
 #include "windowing/X11/XRandR.h"
@@ -158,11 +157,13 @@ bool CWinSystemX11GLES::CreateNewWindow(const CStdString& name, bool fullScreen,
   if(!SetFullScreen(fullScreen, res, false))
 	return false;
 
-  CTexture iconTexture;
-  iconTexture.LoadFromFile("special://xbmc/media/icon256x256.png");
+  CBaseTexture* iconTexture;
+  iconTexture = CTexture::LoadFromFile("special://xbmc/media/icon256x256.png");
 
-  SDL_WM_SetIcon(SDL_CreateRGBSurfaceFrom(iconTexture.GetPixels(), iconTexture.GetWidth(), iconTexture.GetHeight(), BPP, iconTexture.GetPitch(), 0xff0000, 0x00ff00, 0x0000ff, 0xff000000L), NULL);
+  if (iconTexture)
+    SDL_WM_SetIcon(SDL_CreateRGBSurfaceFrom(iconTexture->GetPixels(), iconTexture->GetWidth(), iconTexture->GetHeight(), BPP, iconTexture->GetPitch(), 0xff0000, 0x00ff00, 0x0000ff, 0xff000000L), NULL);
   SDL_WM_SetCaption("XBMC Media Center", NULL);
+  delete iconTexture;
 
   m_bWindowCreated = true;
 
diff --git xbmc-13-0f3db05/xbmc/windowing/X11/WinSystemX11GLES.h xbmc-13-b8667e9/xbmc/windowing/X11/WinSystemX11GLES.h
index b5bd2cc..02ea7de 100644
--- xbmc-13-0f3db05/xbmc/windowing/X11/WinSystemX11GLES.h
+++ xbmc-13-b8667e9/xbmc/windowing/X11/WinSystemX11GLES.h
@@ -25,6 +25,7 @@
 #include "windowing/WinSystem.h"
 #include <EGL/egl.h>
 #include <X11/Xlib.h>
+#include <SDL/SDL.h>
 #include "rendering/gles/RenderSystemGLES.h"
 #include "utils/GlobalsHandling.h"
 
