diff -Nur a/xbmc/cores/dvdplayer/DVDCodecs/Video/Makefile.in b/xbmc/cores/dvdplayer/DVDCodecs/Video/Makefile.in
--- a/xbmc/cores/dvdplayer/DVDCodecs/Video/Makefile.in  2015-01-29 18:47:12.000000000 +0000
+++ b/xbmc/cores/dvdplayer/DVDCodecs/Video/Makefile.in  2015-02-05 20:37:06.930406221 +0000
@@ -26,8 +26,6 @@
 ifeq (@USE_LIBAMCODEC@,1)
 SRCS += AMLCodec.cpp
 SRCS += DVDVideoCodecAmlogic.cpp
-INCLUDES += -I$(prefix)/include/amcodec
-INCLUDES += -I$(prefix)/include/amplayer
 endif

 ifeq (@USE_ANDROID@,1)

