
diff --git a/src/armsoc_driver.c b/src/armsoc_driver.c
index 590352e..3abb80b 100644
--- a/src/armsoc_driver.c
+++ b/src/armsoc_driver.c
@@ -951,7 +951,7 @@ ARMSOCScreenInit(SCREEN_INIT_ARGS_DECL)
 	VisualPtr visual;
 	xf86CrtcConfigPtr xf86_config;
 	int j;
-	char *fbdev;
+	const char *fbdev;
 
 	TRACE_ENTER();
 
diff --git a/src/drmmode_display.c b/src/drmmode_display.c
index b9e59f1..d71b943 100644
--- a/src/drmmode_display.c
+++ b/src/drmmode_display.c
@@ -354,7 +354,7 @@ drmmode_set_mode_major(xf86CrtcPtr crtc, DisplayModePtr mode,
 	drmmode_crtc->last_good_rotation = crtc->rotation;
 	if (drmmode_crtc->last_good_mode) {
 		if (drmmode_crtc->last_good_mode->name)
-			free(drmmode_crtc->last_good_mode->name);
+			free((void *)drmmode_crtc->last_good_mode->name);
 		free(drmmode_crtc->last_good_mode);
 	}
 	drmmode_crtc->last_good_mode = xf86DuplicateMode(&crtc->mode);
