diff -Nur a/include/configs/odroid.h b/include/configs/odroid.h
--- a/include/configs/odroid.h	2015-03-03 20:29:10.000000000 +0000
+++ b/include/configs/odroid.h	2015-03-20 19:11:57.027486019 +0000
@@ -131,6 +131,7 @@
 	"kernel_args=" \
 		"setenv bootargs root=/dev/mmcblk${mmcrootdev}p${mmcrootpart}" \
 		" rootwait ${console} ${opts}\0" \
+	"boot_ini=cfgload\0" \
 	"boot_fit=" \
 		"setenv kerneladdr 0x42000000;" \
 		"setenv kernelname Image.itb;" \
@@ -154,7 +155,9 @@
 		"run kernel_args;" \
 		"bootz ${kerneladdr} ${initrd_addr} ${fdt_addr};\0" \
 	"autoboot=" \
-		"if test -e mmc 0 Image.itb; then; " \
+		"if test -e mmc 0 boot.ini; then; "\
+			"run boot_ini;" \
+		"elif test -e mmc 0 Image.itb; then; " \
 			"run boot_fit;" \
 		"elif test -e mmc 0 zImage; then; " \
 			"run boot_zimg;" \
