diff -Nur a/Makefile b/Makefile
--- a/Makefile	2015-01-06 14:13:59.000000000 -0500
+++ b/Makefile	2015-01-12 12:18:16.409940624 -0500
@@ -374,22 +374,6 @@
 		   -Werror-implicit-function-declaration \
 		   -Wno-format-security \
 		   -fno-delete-null-pointer-checks 
-
-KBUILD_CFLAGS   += -Werror=enum-compare \
-		   -Werror=comment \
-		   -Werror=implicit-int \
-		   -Werror=missing-braces \
-		   -Werror=unused-value \
-		   -Werror=format \
-		   -Werror=switch \
-		   -Werror=strict-prototypes \
-		   -Werror=unused-label \
-		   -Werror=undef \
-		   -Werror=unused-result \
-		   -Werror=return-type \
-		   -Werror=parentheses \
-		   -Werror=int-to-pointer-cast \
-		   -Wno-error=cpp
 KBUILD_AFLAGS_KERNEL :=
 KBUILD_CFLAGS_KERNEL :=
 KBUILD_AFLAGS   := -D__ASSEMBLY__
