--- linux-odroid-u2-3.8.13+0c5ca23/sound/soc/codecs/max98090.c.orig	2014-12-31 20:39:49.018656972 +0000
+++ linux-odroid-u2-3.8.13+0c5ca23/sound/soc/codecs/max98090.c	2014-12-31 20:40:14.578263971 +0000
@@ -2142,7 +2142,7 @@
 EXPORT_SYMBOL_GPL(max98090_mic_detect);
 
 #define MAX98090_RATES SNDRV_PCM_RATE_8000_96000
-#define MAX98090_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE)
+#define MAX98090_FORMATS (SNDRV_PCM_FMTBIT_S16_LE)
 
 static struct snd_soc_dai_ops max98090_dai_ops = {
 	.set_sysclk = max98090_dai_set_sysclk,
